core_impl/src/test/java/com/google/instrumentation/stats/StatsContextTest.java

```diff
@@ -167,17 +169,21 @@ public void testToString() {
   }
 
   private static void testSerialize(Tag... tags) throws IOException {
+    final Map<TagKey, TagValue> tagMap = new HashMap<TagKey, TagValue>();
+    for (Tag tag : tags) {
+      tagMap.put(tag.getKey(), tag.getValue());
+    }
+
     ByteArrayOutputStream actual = new ByteArrayOutputStream();
+    new StatsContextImpl(tagMap).serialize(actual);
+
     ByteArrayOutputStream expected = new ByteArrayOutputStream();
     expected.write(VERSION_ID);
-    StatsContext.Builder builder = DEFAULT.builder();
-    for (Tag tag : tags) {
-      builder.set(tag.getKey(), tag.getValue());
+    for (Entry<TagKey, TagValue> entry : tagMap.entrySet()) {
       expected.write(VALUE_TYPE_STRING);
-      encodeString(tag.getKey().toString(), expected);
-      encodeString(tag.getValue().toString(), expected);
+      encodeString(entry.getKey().asString(), expected);
+      encodeString(entry.getValue().asString(), expected);
     }
-    builder.build().serialize(actual);
 
     assertThat(actual.toByteArray()).isEqualTo(expected.toByteArray());
```

Comment:
That makes sense but I think it would be trivial to generate all the permutations of tags. The key point here is to eliminate the possible inconsistency on tag order between expected and actual, so how about modifying to traverse `statsContextImpl.tags.entrySet()` instead of `tagMap.entrySet()`? In that way even if the constructor copied (or even modified) the map, we could still guarantee the tag orders are the same between expected and actual.